core/test/create-test-trace.js

```diff
@@ -240,7 +246,126 @@ function createTestTrace(options) {
     traceEvents.push(getTopLevelTask({ts: options.traceEnd - 1, duration: 1}));
   }
 
-  return {traceEvents};
+  // TODO(15841): why does "should estimate the FCP & LCP impact" in byte-efficiency-audit-test.js
+  // fail even when not creating records from trace? For now... just don't emit these events for
+  // when using CDP.
+  if (!process.env.INTERNAL_LANTERN_USE_TRACE) {
+    options.networkRecords = undefined;
+  }
+
+  const networkRecords = options.networkRecords || [];
+  for (const record of networkRecords) {
+    // `requestId` is optional in the input test records.
+    const requestId = record.requestId ?
+      record.requestId.replaceAll(':redirect', '') :
+      String(networkRecords.indexOf(record));
+
+    let willBeRedirected = false;
+    if (record.requestId) {
+      const redirectedRequestId = record.requestId + ':redirect';
+      willBeRedirected = networkRecords.some(r => r.requestId === redirectedRequestId);
+    }
+
+    const times = getNormalizedRequestTiming(record);
+    const willSendTime = times.rendererStartTime * 1000;
+    const sendTime = times.networkRequestTime * 1000;
+    const recieveResponseTime = times.responseHeadersEndTime * 1000;
+    const endTime = times.networkEndTime * 1000;
+
+    if (times.timing.receiveHeadersStart === undefined) {
+      times.timing.receiveHeadersStart = times.timing.receiveHeadersEnd;
+    }
+
+    if (!willBeRedirected) {
+      traceEvents.push({
+        name: 'ResourceWillSendRequest',
+        ts: willSendTime,
+        pid,
+        tid,
+        ph: 'I',
+        cat: 'devtools.timeline',
+        dur: 0,
+        args: {
+          data: {
+            requestId,
+            frame: record.frameId,
+            initiator: record.initiator ?? {type: 'other'},
+          },
+        },
+      });
+    }
+
+    traceEvents.push({
+      name: 'ResourceSendRequest',
+      ts: sendTime,
+      pid,
+      tid,
+      ph: 'I',
+      cat: 'devtools.timeline',
+      dur: 0,
+      args: {
+        data: {
+          requestId,
+          frame: record.frameId,
+          priority: record.priority,
+          requestMethod: record.requestMethod,
+          resourceType: record.resourceType ?? 'Document',
+          url: record.url,
+        },
+      },
+    });
+
+    if (willBeRedirected) {
+      continue;
+    }
+
+    traceEvents.push({
+      name: 'ResourceReceiveResponse',
```

Comment:
What about `ResourceReceivedData` / `ResourceChangePriority`? Do we just not need those for this PR?